### ISSUE TITLE
Interrupt blocking IO

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz-zio#75]
Extend the fiber interruption framework to interrupt blocking IO via `Thread.interrupt`.

**Comment by jdegoes** [comment]
The tricky part about this ticket is interrupting the right thread without performance overhead (too much locking, etc.).

**Comment by regiskuckaertz** [comment]
I'd love to help on that one but I need to do some reading first. Assuming we have a way of identifying the right thread, is it just a matter of reading its state and calling `Thread.interrupt` if in either of `BLOCKED` or `WAITING` states?

**Comment by jdegoes** [comment]
The main problem with this ticket is getting the thread the fiber is currently on. This basically requires some kind of locking, because fibers can shift threads. So probably a first start would be `FiberContext.withThread(f: Thread => Unit): Unit`. The `evaluate` loop can then acquire and release a lock (probably a `java.util.concurrent.Semaphore`) right before it starts (to set the thread), and right before it ends (to null out the thread). Then the `withThread` method can acquire the lock, and call the callback (possibly with a null thread if the fiber is not in the evaluate loop). 

After `withThread` is added, then interrupting it will be a matter of calling `withThread(t => if (t != null) interruptThread(t))`, where `interruptThread` can spin in a loop, calling `Thread.interrupt` until the status changes or until it gives up (after a certain number of tries?).

Then testing this would involve creating some blocking IO, e.g. reading unlimited data from std input or something, and then seeing if you can interrupt a fiber that's doing the blocking IO.

**Comment by jdegoes** [comment]
Hmm, some other things to keep in mind:

1. We may want to catch `InterruptedException` and treat it specially inside the `evaluate` loop. Treat it as a normal interruption and reset the interrupt status of the thread.
2. We'll have to be careful not to doubly-interrupt fibers. We need to somehow know when the thread is blocking on IO and it's safe to interrupt that way.
3. We have to respect the interrupt status of the fiber (e.g. the `noInterrupt` count).

**Comment by regiskuckaertz** [comment]
`IO.shift` will be problematic. If we add a `Shift` primitive then it can be handled but otherwise we're out of luck. Something for a follow-up PR perhaps.

**Comment by jdegoes** [comment]
> i.e. only when evaluating SyncEffect or AsyncEffect. Then we can set/release the thread at the edges of these operations (release in resumeAsync in the latter case) and minimise the impact of locking.

This could have severe performance penalties. Ordinarily, evaluation of `SyncEffect` / `AsyncEffect` is super fast (and needs to be). 

We will need some other approach, like monitoring the fiber (from another thread, maybe scheduled) to see if it's making progress and trying to interrupt if it's not making progress and `noInterrupt == 0`. I'm not sure, but we definitely shouldn't lock around the effects since that will cause performance to plummet.

**Comment by regiskuckaertz** [comment]
Aaah yes, I understand. Dang! With a monitoring thread, there is a need for maintaining a representation of the current state (blocked vs not blocked) of each fiber. That is no mean feat, as there can be 1000s of them. Re-locking, would your objection hold with locks that privilege writers unfairly? I'm thinking about [`ReadWriteLock`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReadWriteLock.html) which looks [promising](https://www.ibm.com/developerworks/java/library/j-jtp10264/).

**Comment by regiskuckaertz** [comment]
Alternatively, we could ask the user to make a conscious decision whether to opt-in the `Thread.interrupt` behaviour, for example when the fiber is created. It would be off by default, which implies there is no performance penalty but interruption might block.

**Comment by jdegoes** [comment]
@regiskuckaertz I like that better. Let’s hold off on this ticket for a bit, as I think we need one more first before we can do this right without penalty. I’ll write it up today.

**Comment by ghostdogpr** [comment]
Closed by #536
